cardano-api/src/Cardano/Api/Typed.hs

```diff
@@ -0,0 +1,1142 @@
+                    -> ByteString
+                    -> Either JsonDecodeError a
+deserialiseFromJSON _proxy = either (Left . JsonDecodeError) Right
+                           . Aeson.eitherDecodeStrict'
+
+class SerialiseAsRawBytes a where
+
+  serialiseToRawBytes :: a -> ByteString
+
+  deserialiseFromRawBytes :: AsType a -> ByteString -> Maybe a
+
+serialiseToRawBytesHex :: SerialiseAsRawBytes a => a -> ByteString
+serialiseToRawBytesHex = Base16.encode . serialiseToRawBytes
+
+deserialiseFromRawBytesHex :: SerialiseAsRawBytes a
+                           => AsType a -> ByteString -> Maybe a
+deserialiseFromRawBytesHex proxy hex =
+  case Base16.decode hex of
+    (raw, trailing)
+      | BS.null trailing -> deserialiseFromRawBytes proxy raw
+      | otherwise        -> Nothing
+
+
+-- ----------------------------------------------------------------------------
+-- TextEnvelope Serialisation
+--
+
+type TextEnvelope = TextView.TextView
+type TextEnvelopeType = TextView.TextViewType
+type TextEnvelopeDescr = TextView.TextViewTitle
+
+class (ToCBOR a, FromCBOR a, HasTypeProxy a) => HasTextEnvelope a where
+    textEnvelopeType :: AsType a -> TextEnvelopeType
+
+    textEnvelopeDefaultDescr :: AsType a -> TextEnvelopeDescr
+    textEnvelopeDefaultDescr _ = ""
+
+type TextEnvelopeError = TextView.TextViewError
+
+data FileError e = FileError   FilePath e
+                 | FileIOError FilePath IOException
+
+
+serialiseToTextEnvelope :: forall a. HasTextEnvelope a
+                        => Maybe TextEnvelopeDescr -> a -> TextEnvelope
+serialiseToTextEnvelope mbDescr =
+    TextView.encodeToTextView
+      (textEnvelopeType ttoken)
+      (fromMaybe (textEnvelopeDefaultDescr ttoken) mbDescr)
+      toCBOR
+  where
+    ttoken :: AsType a
+    ttoken = proxyToAsType Proxy
+
+
+deserialiseFromTextEnvelope :: HasTextEnvelope a
+                            => AsType a
+                            -> TextEnvelope
+                            -> Either TextEnvelopeError a
+deserialiseFromTextEnvelope ttoken te = do
+    TextView.expectTextViewOfType (textEnvelopeType ttoken) te
+    TextView.decodeFromTextView fromCBOR te
+
+
+data FromSomeType (c :: * -> Constraint) b where
+     FromSomeType :: c a => AsType a -> (a -> b) -> FromSomeType c b
+
+
+deserialiseFromTextEnvelopeAnyOf :: [FromSomeType HasTextEnvelope b]
+                                 -> TextEnvelope
+                                 -> Either TextEnvelopeError b
+deserialiseFromTextEnvelopeAnyOf types te =
+    case List.find matching types of
+      Nothing ->
+        Left (TextView.TextViewTypeError expectedTypes actualType)
+
+      Just (FromSomeType _ttoken f) ->
+        f <$> TextView.decodeFromTextView fromCBOR te
+  where
+    actualType    = TextView.tvType te
+    expectedTypes = [ textEnvelopeType ttoken
+                    | FromSomeType ttoken _f <- types ]
+
+    matching (FromSomeType ttoken _f) = actualType == textEnvelopeType ttoken
```

Comment:
Okay so this is cool. Given a bunch of ways to convert from some type to `b` and some `TextEnvelope`, this allows us to decode the contents of the `TextEnvelope` and apply a conversion function to that result in order to return a value of type `b`. That is, if we can find an appropriate conversion function from those provided.